### PR TITLE
internal(assistant): Enable autocorrelation with Grafana Assistant

### PR DESCRIPTION
**Pull request by e-fisher**
## Description

- Enable autocorrelation via Grafana Assistant as an alternative AI provider, behind the `grafana-assistant` feature toggle
- Split single `addRule` tool into four selector-specific tools (`addRuleBeginEnd`, `addRuleRegex`, `addRuleJson`, `addRuleHeaderName`) because Grafana Assistant does not support top-level enums in tool schemas. This benefits OpenAI as well (more targeted tool descriptions), so it is not gated behind the feature flag
- Reduce token usage: truncate mismatch values at 80 chars, filter out tracking/infrastructure cookies (Google Analytics, load balancer, session IDs, etc.) from recording data and comparison results
- Make `lastMessageIsToolCall` provider-aware: OpenAI blocks auto-send after `finish` to prevent infinite loop (`toolChoice: 'required'`), Assistant needs a final call to mark conversation as completed
- Move bookkeeping fields (`id`, `type`, `enabled`) out of the AI rule schema; hydrate client-side in `toCorrelationRule`
- Hide `TokenUsageIndicator` for Assistant, will be removed once we switch
- Basic error handling for Grafana Assistant (session expiry, not signed in, generic error). Follow-up PR will have improved error handling.

## How to Test
- With `grafana-assistant` feature toggle ON: run autocorrelation end-to-end, verify it solves same correlation challenges as OpenAI
- With feature toggle OFF: verify OpenAI autocorrelation works as before (rules created, validation runs, finish outcome displayed)

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added tests for my changes.
- [x] I have commented on my code, particularly in hard-to-understand areas.

## Related PR(s)/Issue(s)

---

> [!NOTE]
> **Medium Risk**
> Touches AI streaming/tool-call completion logic and introduces a new provider path (A2A), where subtle ordering/continuation edge cases could cause hangs or premature finishes. Schema/tooling changes also affect rule creation across providers, though changes are covered by updated and new tests.
> 
> **Overview**
> Enables autocorrelation to run via **Grafana Assistant (A2A)** behind the `grafana-assistant` feature toggle, including provider-aware auto tool continuation and UI adjustments (hide token usage for Assistant, new reset flow, and Assistant-specific error states like session expiry/not signed in).
> 
> Updates the A2A streaming/tool-call flow to avoid premature `finish(tool-calls)` by gating on `step.complete(tool_use)` (`allToolCallsReceived`) and adds coverage for fragmented/interleaved tool calls across continuation streams.
> 
> Splits the single `addRule` tool into selector-specific tools (`addRuleBeginEnd`, `addRuleRegex`, `addRuleJson`, `addRuleHeaderName`) with updated schemas/types, and moves rule bookkeeping (`id`, `type`, `enabled`) to client-side hydration. Separately reduces token usage/noise by filtering common infra/analytics cookies from AI inputs and header comparisons, and truncating mismatch values in validation output.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9e155f9abe1fac66e5f0d5aadc43ff7ab441da13. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by allansson — COMMENTED** [review]
There seems to be a race condition of some kind. Sometimes when I run it, it will stop at the "Create rules" step and the logs say "No pending request for toolCallId=toolu_bdrk_015ZxDBQpv6gy5tYv6CZ2auQ. Skipping."

Image

I also got this error:

image

Not sure what to do with that. It seems like it could be something in Grafana Assistant?

**Comment by e-fisher** [comment]
Good catch! Race condition in A2A stream, should be fixed now.

Permission check failed fixed in #1171

File: src/handlers/ai/tools.ts [hunk 1]
```diff
@@ -82,10 +101,24 @@ export const tools = {
...
+  }),
+
+  addRuleJson: tool({
+    description: 'Create a correlation rule with a JSON-path selector.',
+    inputSchema: z.object({ rule: addRuleJsonSchema }),
+  }),
+
+  addRuleHeaderName: tool({
+    description: 'Create a correlation rule with a header-name selector.',
+    inputSchema: z.object({ rule: addRuleHeaderNameSchema }),
```
[the remark this answers is not in the document]
Comment (e-fisher) [reply src/handlers/ai/tools.ts:121]: non issue

File: src/handlers/ai/grafanaAssistantProvider.ts [hunk 2]
```diff
@@ -144,7 +144,7 @@ export class GrafanaAssistantLanguageModel implements LanguageModelV2 {
   ): Promise<Awaited<ReturnType<LanguageModelV2['doStream']>>> {
     forwardAbortSignal(abortSignal, session.sessionAbortController)
 
-    session.readyToFinishForTools = false
+    session.allToolCallsReceived = false
```
[the remark this answers is not in the document]
Comment (e-fisher) [reply src/handlers/ai/grafanaAssistantProvider.ts]: Fixed by converting readyToFinishForTools into getter

**Review by allansson — APPROVED** [review]
Tested both GA and OpenAI. Both produce something working...

Great job! 🥳
